File: Go/matrices/diagonalMatrix.go
package matrices

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

// Diagonal matrix structure.
type DiagonalMatrix struct {
	// Size of diagonal matrix.
	Size int
	// Diagonal matrix itself.
	Diag []float64
}

// Creating an instance of a diagonal matrix and returning a pointer.
func NewDiagonalMatrix(size int) *DiagonalMatrix {
	dm := new(DiagonalMatrix)
	dm.Size = size
	dm.Diag = make([]float64, size)
	return dm
}

// File input.
func (dm *DiagonalMatrix) In(line string) {
	strs := strings.Split(line, " ")
	for i := 0; i < dm.Size; i++ {
		dm.Diag[i], _ = strconv.ParseFloat(strs[i], 64)
	}
}

// Random input.
func (dm *DiagonalMatrix) RandomIn() {
	for i := 0; i < dm.Size; i++ {
		dm.Diag[i] = float64(rand.Int31()%100) * rand.Float64()
	}
}

// File output.
func (dm *DiagonalMatrix) Out(f *os.File) {
	f.WriteString("Size: ")
	f.WriteString(fmt.Sprintf("%d\n", dm.Size))
	f.WriteString("Average: ")
	f.WriteString(fmt.Sprintf("%f\n", dm.GetAverage()))
	f.WriteString("Type: Diagonal Matrix\n")

	for i := 0; i < dm.Size; i++ {
		for j := 0; j < dm.Size; j++ {
			// Only diagonal elements can be non-zero.
			if i == j {
				f.WriteString(fmt.Sprintf("%f ", dm.Diag[i]))
			} else {
				f.WriteString("0 ")
			}
		}
		f.WriteString("\n")
	}
}

// Getting average of all the elements.
func (dm *DiagonalMatrix) GetAverage() float64 {
	// Total sum of all the elements.
	var sum float64 = 0

	for i := 0; i < dm.Size; i++ {
		sum += dm.Diag[i]
	}

	return sum / float64(dm.Size*dm.Size)
}

File: Python/matrices/diagonalMatrix.py
from matrices.baseMatrix import BaseMatrix
import random


# Diagonal matrix class.
class DiagonalMatrix(BaseMatrix):
    # Diagonal matrix size.
    __size__: int
    # Diagonal matrix itself.
    __diag__: list

    # Diagonal matrix constructor.
    def __init__(self, size: int):
        super().__init__(size)
        self.__diag__ = []

    # Reading from file.
    def read(self, line: str):
        # Filling diagonal matrix with numbers.
        self.__diag__ = list(map(float, line.split()))

    def random_read(self):
        self.__size__ = random.randint(0, 100) + 1
        self.__diag__ = []

        # Filling diagonal matrix with numbers.
        for i in range(0, self.__size__):
            self.__diag__.append(random.random() * random.randint(0, 1000))

    # Outputting.
    def out(self):
        ans = 'Size: ' + str(self.__size__) + '\n'
        ans += 'Average: ' + str(self.get_average()) + '\n'
        ans += 'Diagonal matrix:\n'

        for i in range(0, self.__size__):
            for j in range(0, self.__size__):
                if i == j:
                    ans += str(self.__diag__[i]) + ' '
                else:
                    ans += '0 '
            ans += '\n'

        return ans

    # Getting average.
    def get_average(self) -> float:
        # Total sum of elements.
        summ = 0

        for i in range(0, self.__size__):
            summ = summ + self.__diag__[i]

        return summ / self.__size__ / self.__size__
